Components: add dismissable-card

gwwar opened the pull request:
As part of #7329 This PR adds a new block component. This is a card that remembers when it's been dismissed. You can configure this to dismiss it for the current page load, or forever (saved as a user preference).

![dismiss](https://cloud.githubusercontent.com/assets/1270189/18151403/d0a6c66c-6fa3-11e6-9f7c-eb8600d383db.gif)

## Testing Instructions
- Navigate to http://calypso.localhost:3000/devdocs/blocks/dismissable-card
- You can dismiss both cards
- Refreshing the page will render only the top card
- No other functional changes since there are no usages yet in Calypso

cc @lamosty @artpi @aduth 

Test live: https://calypso.live/?branch=add/dismissable-card

lamosty commented:
I know it's still in progress but reviewed anyway. :)

Nice work! I learned something new from the review. I only got three things for now:

1. After hovering over the X (dismiss) icon, we need to change the cursor to pointer;
2. I think we could store the preferences a bit better: instead of storing them as a key: value in the main `calypso_preferences` obj, what if we added a `dismissable-cards` obj with keys being the `preferenceName`s and values `true` or `false`. This way, even if we got a lot of dismissable cards in use, we don't pollute the global `calypso_preferences` namespace. What do you think?
3. Why is the key called `dismissable-cards` when all the other keys (except `editor-mode` and some camelCase ones) use snake_case?

mtias commented on assets/stylesheets/_components.scss:
```diff
@@ -10,9 +10,10 @@
 @import 'blocks/comment-button/style';
 @import 'blocks/comments/style';
-@import 'blocks/author-selector/style';
+@import 'blocks/dismissable-card/style';
```
This is an interesting one. Given it's fairly generic I could expect it to be in `components` but the connection with user preferences throws a bit of wrench. :)

aduth commented on client/blocks/dismissable-card/docs/example.jsx:
```diff
@@ -0,0 +1,49 @@
+}
+
+const ConnectedDismissableCardExample = connect( 
+	( state ) => { return {}; },
```
You can pass `null` as the argument if you don't want to select anything. This could also be expressed with an implicit return, i.e.: `( state ) => ( {} ),`

gwwar replied:
Cleaned up in 53444d3

aduth commented on client/blocks/dismissable-card/docs/example.jsx:
```diff
@@ -0,0 +1,49 @@
+
+const ConnectedDismissableCardExample = connect( 
+	( state ) => { return {}; },
+	( dispatch ) => bindActionCreators( {
```
Not sure it's any better, but you could use [`_.partial`](https://lodash.com/docs/4.15.0#partial) to generate the bound function (or `Function#bind` for that matter), allowing you to use the object notation for `mapDispatchToProps`:

```js
{ clearPreference: partial( savePreference, 'dismissable-card-example, 'null ) }
```

gwwar replied:
It does look tidier, added in 53444d3

aduth commented on client/blocks/dismissable-card/README.md:
```diff
@@ -0,0 +1,67 @@
+render: function() {
+  return (
+    <div className="your-stuff">
+      <QueryPreferences />
```
I'm not sure what we should do about `<QueryPreferences />`. While in the traditional sense, query components should be rendered this way to ensure that the data is kept fresh in response to the dependent component's own lifecycle, we already [render a `<QueryPreferences />` in `<Layout />`](https://github.com/Automattic/wp-calypso/blob/6bb2e62f614bd6d58828b27e8db87a36d4482062/client/layout/index.jsx#L201) so we can be guaranteed that the request will always be made on every page load.

mtias replied:
Relying on the presence of QueryPreferences up the tree seems like a strange coupling to do. Though I think the QueryPreferences should be placed within the dismissable-card component if possible.

gwwar replied:
I went ahead and moved it into `DismissableCard` in 7a4b1e5

gwwar commented:
> After hovering over the X (dismiss) icon, we need to change the cursor to pointer;

Nice catch, updated in c677e9d

> instead of storing them as a key: value in the main calypso_preferences obj, what if we added a dismissable-cards obj

I was hoping to keep the types simple. We might run into stale data updates if folks happen to be working in two tabs. If people have strong feeling to update this, I can.

> Why is the key called dismissable-cards when all the other keys (except editor-mode and some camelCase ones) use snake_case?

I didn't really see a winning pattern, and I tend to name settings in `kebab-case` vs `snake_case` or `camelCase`. If we do have a preference for another type I can switch to `camelCase` or `snake_case`.

lamosty commented:
> I was hoping to keep the types simple. We might run into stale data updates if folks happen to be working in two tabs. If people have strong feeling to update this, I can.

Good point, I see. Let's keep it simple, then.

The dismissable card works fine with more content as well:

![selection_147](https://cloud.githubusercontent.com/assets/4988512/18249897/e3b4c6a6-7380-11e6-91f7-d11c81711630.png)

The 'X' button stays in the top right corner, nice.

LGTM :ship:

gwwar commented:
Thanks for the review @lamosty! I'll rebase and test again before 🚢
